### PR TITLE
Performance patches prerequisite

### PR DESCRIPTION
**Pull request by Spottedleaf**
Original patches:
https://github.com/Spottedleaf/Concrete/blob/bleeding/patches/server/0004-Util-patch.patch
https://github.com/Spottedleaf/Concrete/blob/bleeding/patches/server/0022-Improve-autosave-performance.patch
https://github.com/Spottedleaf/Concrete/blob/bleeding/patches/server/0023-Avoid-immediately-autosaving-newly-generated-loaded-.patch

See patch notes for details
The patches aren't always 1:1 ports, they may have been changed to improve behavior or performance in some areas

They've all been dumped in this PR since they tend to have dependencies on each other (most notably the mc-util patch)

### CODE REVIEW

**Comment by electronicboy** [comment]
pre-looking at the patches comment: gfdi leaf

**Comment by Spottedleaf** [comment]
now it's just the patches that cause big diffs

**Comment by AJMFactsheets** [comment]
I built this branch for local testing and the only problem I found is that when a player dies, there is a high chance they will get stuck in a limbo mode where chunks won't load, entities jitter, and they player can't move.

**Comment by Spottedleaf** [comment]
can't produce locally, paste output of /version

**Comment by HexedHero** [comment]
#LeafForProjectDev

**Comment by Callahhh** [comment]
FWIW, here's a spark report with ~35 players online:
 https://sparkprofiler.github.io/#Oes5zdvftS. 
This is one without the patches:
https://sparkprofiler.github.io/#msNHaXfBlD

Here's also a timings report with the patches (30ish minutes):
https://timings.aikar.co/?id=a872c4b888c54ed2bb067fc7c2b43c11

I had to revert back since TPS was worse than usual. After looking/comparing both of those reports for like an hour, I couldn't find anything noticeably different or where it went wrong. 

I apologize if this is all pretty worthless info, just thought I'd share. 

Appreciate the hard work ❤️

**Comment by electronicboy** [comment]
mostly looks like your server is overworked from mob spawning/player movement

**Comment by Spottedleaf** [comment]
yeah I don't see much of a difference between the two, and without a testing environment I can't do anything about it

**Comment by AJMFactsheets** [comment]
Output of /version: This server is running Paper version git-Paper-"59df758f" (MC: 1.15.1) (Implementing API version 1.15.1-R0.1-SNAPSHOT)

After further testing, I have found that the death bug only happens after a player has slept in a bed.

**Comment by Spottedleaf** [comment]
unable to produce after sleeping in a bed

**Comment by AJMFactsheets** [comment]
I built a new jar based off of the new commit and the death glitch still persists. I realized beds have nothing to do with the glitch, but it has something to do with spawn chunks.

Start a new world with default settings, and /kill yourself at the world spawn. Upon respawning, you should be unable to move, unless you relog or teleport far away.

The glitch reminds me of [MC-138539](https://bugs.mojang.com/browse/MC-138539)
I also tested with 1.15.1 Paper-31 and didn't have any problems.

**Comment by Spottedleaf** [comment]
Yeah I see what's going on here now - a typo. The reason it only happens at spawn is because once chunks get loaded they are sent to the player - at spawn they only load once. 

Fixed in latest commit.

Thanks for your patience and detailed report.

**Review by electronicboy — COMMENTED** [review]
- [x] Spottedleaf/Concrete:patches/server/0004-Util-patch.patch@bleeding 

- [ ] Spottedleaf/Concrete:patches/server/0022-Improve-autosave-performance.patch@bleeding 
- [x] Spottedleaf/Concrete:patches/server/0023-Avoid-immediately-autosaving-newly-generated-loaded-.patch@bleeding 
- [ ] Spottedleaf/Concrete:patches/server/0013-Per-player-view-distance-implementation.patch@bleeding 
- [ ] Spottedleaf/Concrete:patches/server/0015-Optimize-isOutsideRange-to-use-a-square-radius-check.patch@bleeding 
- [x] Spigot-Server-Patches/0420-Optimise-getChunkAt-calls-for-loaded-chunks.patch

File: scripts/importmcdev.sh [hunk 1]
```diff
@@ -106,6 +106,7 @@ done
 
 # dont forget \ at end of each line but last
 importLibrary com.mojang authlib com/mojang/authlib yggdrasil/YggdrasilGameProfileRepository.java
+importLibrary com.mojang datafixerupper com/mojang/datafixers/util Either.java
```
Comment (electronicboy) [inline scripts/importmcdev.sh:109]: Either?!

**Review by electronicboy — APPROVED** [review]

**Comment by Spottedleaf** [comment]
updated to 1.15.2, fixed build

**Review by electronicboy — CHANGES_REQUESTED** [review]
https://www.youtube.com/watch?v=woySeSNBL3o

**Review by electronicboy — APPROVED** [review]

**Review by zachbr — APPROVED** [review]

**Comment by Misio12320** [comment]
How much % the game has been upgraded with performance and when release of view distance for 1.15.2?

**Comment by perseus5** [comment]
> How much % the game has been upgraded with performance and when release of view distance for 1.15.2?

yes

**Comment by Spottedleaf** [comment]
it's a prereq patch so like maybe 2% at most

**Comment by jonttm** [comment]
Are you going to bring the per-player view distance patch back in the future?

**Comment by Spottedleaf** [comment]
yes, some other stuff needs to go in first though

**Comment by NicoNekoDev** [comment]
> it's a prereq patch so like maybe 2% at most

that's a lot :O gonna test it now on my 20gb server, i'll report back the performance.
Edit: here's is the performance report: https://timings.aikar.co/?id=8fe23d8d5ecf4518a4c5d132e73e213a (i've maxed to 32 gb)
The only thing I've noticed was the chunks loading in a V form when I travel fast.
My only problem is the new chunks beeing generated very slow, also it looks like I do have somewhere a memory leak, as the server ram start to increase over time.

**Pull request merged**